tests/ctest/test_keyset_cow.c

```diff
@@ -0,0 +1,349 @@
+
+	keyDel (root);
+	ksDel (ks);
+	ksDel (cut);
+}
+
+static void ksCut_cascading_key_copied_from_root_in_keyset_should_work (void)
+{
+	printf("Test %s\n", __func__);
+
+	// Arrange
+	Key * root = keyNew ("/root", KS_END);
+	Key * copy = keyCopy (keyNew ("/", KEY_END), root, KEY_CP_ALL);
+
+	KeySet * ks = ksNew (KEYSET_SIZE, copy, keyNew ("system:/root/abc", KEY_END), keyNew ("system:/other", KEY_END), KS_END);
+
+	// Act
+	KeySet * cut = ksCut (ks, root);
+
+	// Assert
+	succeed_if_fmt(ksGetSize (cut) == 2, "should contain 2 keys, was %zd", ksGetSize (cut));
+	succeed_if (ksLookupByName (cut, "/root", 0) != NULL, "should contain root");
+	succeed_if (ksLookupByName (cut, "system:/root/abc", 0) != NULL, "should contain key below root");
+	succeed_if (ksGetSize (ks) == 1, "only 1 key should remain in keyset");
+	succeed_if (ksLookupByName (ks, "system:/other", 0) != NULL, "ks should contain other key");
+
+	keyDel (root);
+	keyDel (copy);
+	ksDel (ks);
+	ksDel (cut);
+}
+
+
+
+static void ksFindHierarchy_root_in_keyset_should_work (void)
+{
+	printf("Test %s\n", __func__);
+
+	// Arrange
+	Key * root = keyNew ("system:/root", KS_END);
+
+	KeySet * ks = ksNew (KEYSET_SIZE, root, keyNew ("system:/root/abc", KEY_END), keyNew ("system:/other", KEY_END), KS_END);
+
+	elektraCursor end = 999;
+
+	// Act
+	elektraCursor start = ksFindHierarchy (ks, root, &end);
+
+	// Assert
+	// keys in the keyset are ordered alphabetically.
+	// - system:/other
+	// - system:/root
+	// - system:/root/abc
+
+	succeed_if_fmt(start == 1, "start should be 1, was %zd", start);
+	succeed_if_fmt(end == 3, "end should be 3, was %zd", end);
+
+	keyDel (root);
+	ksDel (ks);
+}
+
+static void ksFindHierarchy_key_copied_from_root_in_keyset_should_work (void)
+{
+	printf("Test %s\n", __func__);
+
+	// Arrange
+	Key * root = keyNew ("system:/root", KS_END);
+	Key * copy = keyCopy (keyNew ("/", KEY_END), root, KEY_CP_ALL);
+
+	KeySet * ks = ksNew (KEYSET_SIZE, copy, keyNew ("system:/root/abc", KEY_END), keyNew ("system:/other", KEY_END), KS_END);
+
+	elektraCursor end = 999;
+
+	// Act
+	elektraCursor start = ksFindHierarchy (ks, root, &end);
+
+	// Assert
+	// keys in the keyset are ordered alphabetically.
+	// - system:/other
+	// - system:/root
+	// - system:/root/abc
+
+	succeed_if_fmt(start == 1, "start should be 1, was %zd", start);
+	succeed_if_fmt(end == 3, "end should be 3, was %zd", end);
+
+	keyDel (root);
+	keyDel (copy);
+	ksDel (ks);
+}
+
+static void ksFindHierarchy_cascading_root_in_keyset_should_work (void)
+{
+	printf("Test %s\n", __func__);
+
+	// Arrange
+	Key * root = keyNew ("/root", KS_END);
+
+	KeySet * ks = ksNew (KEYSET_SIZE, root, keyNew ("system:/root/abc", KEY_END), keyNew ("system:/other", KEY_END), KS_END);
+
+	elektraCursor end = 999;
+
+	// Act
+	elektraCursor start = ksFindHierarchy (ks, root, &end);
+
+	// Assert
+	// keys in the keyset are ordered alphabetically.
+	// - system:/other
+	// - /root
+	// - system:/root/abc
```

Comment:
That order would be wrong. `/root` should be first. Keys are ordered by namespace according to 
https://github.com/ElektraInitiative/libelektra/blob/4283744cc4689792a8143d55f5f49e6dc1eac81c/src/include/kdb.h.in#L86-L96

Reply:
Okay, but does that mean that a hierarchy, as defined by `ksFindHierarchy`, is defined by the namespace of the keys?

So 

- `system:/a`
- `system:/a/b`

are a different hierarchy from

- `user:/a`
- `user:/a/b`

?

Reply:
Yes, the range returned by `ksFindHierarchy` will always contain keys of a single namespace only and that namespace must be the same as the one used by `root`.

It's obvious for the normal namespaces `system:/`, `user:/` etc., because by any definition `user:/` key is never below a `system:/` key or vice versa. But even if `root` uses `KEY_NS_CASCADING`, only cascading keys will be selected, so `ksFindHierarchy` uses a stricter definition of "below" than e.g. `keyIsBelow`.

If you think about it, there is no other way of doing this. Because we return only a `start` and `end` index, we can only select a continuous range. How would that work with `root = /a` for e.g.

```
user:/a/b
user:/a/b/c
user:/b/c
system:/a/b
```

If you want `user:/a/b` and `system:/a/b` to be selected, you have to select `user:/b/c` as well, because of the order.

We could change the order to sort by namespace after the rest of the name, i.e.:

```
user:/a/b
system:/a/b
user:/a/b/c
user:/b/c
```

But this now has a problem with `root = user:/a` where it is very obvious what that the result should always be:

```
user:/a/b
user:/a/b/c
```

And just to clarify, you can't sort `user:/a/b/c` before `system:/a/b`, because then we are back to the first problem, but now with `root = /a/b`

```
user:/a/b
user:/a/b/c
system:/a/b
user:/b/c
```

> **Note** The true reason is actually just that the namespace is stored as the first byte of the unescaped name and the order of keys is determined by a simple `memcmp` on the whole unescaped name. But as you can see, the current behaviour is still the best option, even if we'd use another way to sort keys.

Reply:
Thanks for the detailed writeup! Will add some of it to the documentation of `ksFindHierarchy`.